### PR TITLE
Add basic metrics

### PR DESCRIPTION
**Pull request by alban-auzeill**

### CODE REVIEW

File: common-rule-engine/src/main/java/org/sonar/commonruleengine/MetricsVisitor.java [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+  }
+
+  public void visitNode(UastNode node) {
+    if (node.kinds.contains(UastNode.Kind.CLASS)) {
+      metrics.numberOfClasses++;
+    }
+    if (node.kinds.contains(UastNode.Kind.FUNCTION)) {
+      metrics.numberOfFunctions++;
+    }
+    if (node.kinds.contains(UastNode.Kind.STATEMENT)) {
```
Comment (ghost) [inline common-rule-engine/src/main/java/org/sonar/commonruleengine/MetricsVisitor.java:21]: Could it ever make sense in any language that a node has 2 or more kinds of CLASS, FUNCTION and STATEMENT at the same time? If we want to support this, then I think we should have unit tests for it.

File: common-rule-engine/src/main/java/org/sonar/uast/UastNode.java [hunk 2]
```diff
@@ -23,14 +24,42 @@ public UastNode(Set<Kind> kinds, String nativeNode, @Nullable Token token, List<
...
+    public final int endLine;
+    /**
+     * start at 1, column number of the last character of the token
+     * if the token has only one character, endColumn() == column()
+     * EOF token is empty, in this case endColumn() == column() - 1
+     */
+    public final int endColumn;
 
     public Token(int line, int column, String value) {
       this.line = line;
```
Comment (ghost) [inline common-rule-engine/src/main/java/org/sonar/uast/UastNode.java]: Is it ok to assume valid `line` and `column` values? (>= 1)

**Review by ghost — APPROVED** [review]
LGTM. Any ideas why `NoSelfAssignmentCheckTest.java` shows up in the diff with no changes?

File: common-rule-engine/src/main/java/org/sonar/commonruleengine/MetricsVisitor.java (hunk 1 above)
Comment (alban-auzeill) [reply common-rule-engine/src/main/java/org/sonar/commonruleengine/MetricsVisitor.java:21]: I added a test

File: common-rule-engine/src/main/java/org/sonar/uast/UastNode.java (hunk 2 above)
Comment (alban-auzeill) [reply common-rule-engine/src/main/java/org/sonar/uast/UastNode.java]: Now line and column are mandatory.

**Pull request merged**